### PR TITLE
refactor open method and the file object

### PR DESCRIPTION
**Pull request by liormizr**

### CODE REVIEW

**Comment by LeoQuote** [comment]
I like this idea that we can finally get bucket name as str , so I can use it directly.

I think it can also fix the exception when opening file with encoding parameter.

```
  File "D:\doubancodes\bandersnatch\venv\lib\site-packages\s3path.py", line 213, in open
    return file_object(
  File "D:\doubancodes\bandersnatch\venv\lib\site-packages\s3path.py", line 769, in __init__
    self._cache = NamedTemporaryFile(
  File "D:\sdks\Python3.8.8\lib\tempfile.py", line 542, in NamedTemporaryFile
    file = _io.open(fd, mode, buffering=buffering,
ValueError: binary mode doesn't take an encoding argument
```

**Comment by cooperlees** [comment]
Hi @liormizr - I'm the maintainer of bandersnatch (PyPI Mirror software) and was wondering what's left to get this ready for merge? Is there anything I or @LeoQuote can help with? This will tremendously help @liormizr adding a S3 storage plugin for our mirroring software.

Thanks!

**Comment by LeoQuote** [comment]
this pr seems to be under development, It cannot work on customized endpoint and access_key , etc...

things could be fixed if we use predefined boto3 session instead of `boto3.DEFAULT_SESSION`

File: s3path.py [hunk 1]
```diff
@@ -205,23 +180,57 @@ def listdir(self, path):
...
             mode=mode,
             buffering=buffering,
             encoding=encoding,
             errors=errors,
-            newline=newline)
+            newline=newline,
+            ignore_ext=True,
+            transport_params={
+                'session': boto3.DEFAULT_SESSION,
+                'resource_kwargs': get_resource_kwargs(),
```
Comment (LeoQuote) [inline s3path.py:223]:
```python
        s3path_session = boto3.Session()
        s3path_client = s3path_session.client(
            's3',
            endpoint_url=resource_kwargs['endpoint_url'],
            aws_access_key_id=resource_kwargs['aws_access_key_id'],
            aws_secret_access_key=resource_kwargs['aws_secret_access_key'],
            aws_session_token=resource_kwargs['aws_session_token']
        )
        file_object = smart_open.open(
            uri=path.as_uri(),
            mode=mode,
            buffering=buffering,
            encoding=encoding,
            errors=errors,
            newline=newline,
            ignore_ext=True,
            transport_params={
                'session': s3path_session,
                'client': s3path_client,
                'resource_kwargs': resource_kwargs,
```

resource_kwargs seems not working when using custom endpoint , I set it on transport_params['client'] and transport_params['session'] instead and it worked ! 

please consider my method , this is also on offical readme page.

Comment (liormizr) [reply s3path.py:223]: Hi @LeoQuote and @cooperlees sorry for the delay in response.

Great catch!
Let's add your fix as well.

Regarding timeline, I'll start writing the changelog for this big version.
After that you will have a new version. :-)

ETA - two days +-

Comment (cooperlees) [reply s3path.py:223]: Awesome. Many thanks. Let’s us know if we can help. We’ll be testing with bandersnatch soon as released.

**Pull request merged**